Improvements to evaluation user interface

skulonen opened the pull request:
# Description

**What?**

This PR improves the submission inspection and assessment UI.

**Why?**

We received feedback from teachers about usability issues in the submission inspection and assessment UI. This PR aims to fix those issues.

**How?**

The inspection and assessment pages have been combined into one page with a completely redesigned user interface. The redesign fixes some of the issues:
- The UI is less cluttered: technical details are now hidden by default but can still be displayed by clicking a button.
- Submitted files can now be viewed more easily instead of opening them in a popup window. The page has a two-column layout in which submitted files are always displayed in the left column.

When assessing a submission, there is now an option to mark the given points as final. When this option is activated, the points given by the assesser override the points from any other submission, even if the other submissions have more points. This affects all views where the best/final score of the exercise is displayed: the course main page, exercise page, API, etc.

In the submission list, there is now an option to group the submissions by submitter, instead of listing all submissions separately. In addition, there is a "start manual assessment" button, which directly takes the user to the first submitter that has not yet been assessed manually. On the inspection/assessment page there is a "next unassessed submitter" link, which takes the user to the next submitter after the currently displayed submitter that has not yet been assessed. This improves the workflow of assessing all submitters in a row. Before this PR, users had to go through the submission list and manually find unassessed submitters, and open them in new tabs to assess them.

The course sidebar can now be collapsed. This is especially useful in the assessment page since it allows more screen area for the main content of the page.

Fixes #831


# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)


**What type of test did you run?**

- [x] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [x] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

There are several accessibility issues reported by WAVE, but none of them are caused by these changes: low contrast in the highlight.js code blocks, label-less invisible textareas next to the "copy to clipboard" button, etc.

**Did you test the changes in**

- [x] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [x] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

Mankro commented on exercise/cache/points.py:
```diff
@@ -113,23 +123,36 @@ def r_augment(children):
-                        'unofficial': unofficial,
+                        'passed': submission.grade >= entry['points_to_pass'],
+                        'graded': True,
+                        'unofficial': False,
```
What happens if the teacher tries to mark an unofficial submission as the "final" one in the manual assessment UI? Did you test that or think about it?

skulonen replied:
I didn't think about that. Looks like it makes the UI a bit confusing: the points appear to be color coded as grey in some places, and not grey in other places. For example, submission 4 is marked as final here:
![image](https://user-images.githubusercontent.com/28843787/136018869-f1d0cadd-d968-49d1-87be-3daa2213e690.png)
I think there are 3 options to solve this:
1. Marking a submission as final changes its status from 'unofficial' to 'ready'. I think this would make sense because the label for 'unofficial' is 'no effect on grading', but that is not true when a submission marked as final: it is literally the only thing that has any effect an grading.
2. Can't mark unofficial submissions as final.
3. Display unofficial submissions, that are marked as final, consistently as official everywhere in the UI, despite being stored as unofficial submissions.

We _could_ also do the reverse of option 3: display them as unofficial everywhere in the UI. But that wouldn't really make sense, because the definitions of 'unofficial' and 'final' are conflicting, as explained in point 1.

Mankro replied:
Option 1 makes a lot of sense. If a teacher wants to turn an unofficial submission into a graded submission, they could do it by marking the submission as final. (There are use cases for that.) The submission status can be then changed to "ready" from "unofficial".

Please test that this works well. We haven't changed unofficial submissions to graded submissions in the past. Pay attention to the student's and teacher's UI, points cache, all results table and the points API.
In addition, does it matter if the student has several normal and unofficial submissions? There could be a few normal submissions, a few unofficial submissions. If one unofficial submission is turned to a normal graded submission and there are unofficial submissions between the normal submissions, I hope that doesn't cause any issues.

Mankro replied:
Marking a submission as final should always set the submission status to "ready". If we manually assess a submission that is stuck in the statuses "initialized" or "waiting for grading" and mark it as final, then the status should be "ready" after that.

Mankro replied:
If a teacher manually assesses as submission and marks it as final, then is it still possible that the automatic grading overrides the points if the exercise service sends the points after the manual assessment?

skulonen replied:
Good points. I'll implement option 1 (submissions are always set to status 'ready' when marking them as final) and test it throughly.

> If a teacher manually assesses as submission and marks it as final, then is it still possible that the automatic grading overrides the points if the exercise service sends the points after the manual assessment?

Yes, this will happen. When points are received from the grader, the status is always set to 'ready', unless it was 'unofficial'. We probably don't want that to happen, and instead the points given by the teacher should remain final.

Another thing that came to my mind: what if the teacher marks one unofficial submission as final, then decides that another unofficial submission should be graded instead, and marks that one as final? In that case, the first submission is no longer final, but it won't be unofficial either - it will remain in the 'ready' state. This doesn't affect grading, because the second submission is now final, but the original submission has still changed its state, with no way to change it back (apart from the admin UI). We can't really change it back automatically when the second submission is marked as final, because we don't store the previous state of the first submission before it was marked as final.

Mankro replied:
**Automatic grading after manual assessment:** this is tricky since sometimes teachers might want to regrade submissions automatically after writing manual feedback. The database stores only one grade and we don't really know if it has been set automatically or manually. I would say we could keep this simple at first and ignore this, in other words, allow automatic grading to override the grade. It should be rare that people somehow accidentally auto-grade submissions after manual assessment (so it should be intentional when auto-grading really happens). Usually auto-grading has finished before manual assessment is started and the manual grade is then anyway the last one. There might also be situations when teachers want to mark one submission as final, but they want to use the auto-grading results without any manual assessment (especially if they want to convert unofficial submissions to graded submissions).

**Multiple unofficial submissions as final:** let's assume there is no problem in leaving the first unofficial submission in the ready state (which makes it official). If it doesn't break anything, then it's OK. Since unofficial submissions exist, the student has already lost the opportunity to make new graded submissions. Therefore, the teacher is trying to fix something manually by marking submissions as final and the relevant outcome is that the selected submission really defines the final grade.

skulonen replied:
Now the submission is set to status 'ready' when marking it as final. I tested and did not find any issues caused by this in the student UI, teacher UI, points cache, all results page or the API.

Mankro commented on exercise/exercise_summary.py:
```diff
@@ -36,27 +39,46 @@ def __init__(self, exercise, user=None):
+                        # This submission is chosen as the best submission and
+                        # no further submissions are considered.
                         self.best_submission = s
                         self.unofficial = False
```
Like in exercise/cache/points.py, what if an unofficial submission is marked as the final graded submission in the manual assesment UI?

skulonen replied:
See previous comments, submissions are now set to status 'ready' when they are marked as final.

Mankro reviewed (changes requested):
Overall, this is very good! I have written comments in most files. Some small fixes are needed. The following list includes issues that repeat multiple times:

* perhaps later in the autumn, we could refactor the points code so that similar logic is not repeated in many places (DRY principle), but not now due to time constraints.
* you didn't use `const` in JavaScript code, but it would be preferred whenever the variable is not supposed to be reassigned. Perhaps it is not worth the time to go over all your JS code now, though?
* accessibility concerns: `<a>` should be used for links that take the user to another page and `<button>` for actions that modify the current page, such as hide the course navigation. Using `<button>` also avoids using the misleading `href="#"` in fake links. When a link is a real link that leads to another page, it should have the default role link and not use the `role="button"` that you have in many links. I have suggested removing the role button from most of the `<a>` links in your code.
* The Bootstrap html class `hide` is deprecated since BS v3.01. The class `hidden` should be a direct replacement. You have used "hide" a lot, but changing them to "hidden" is easy, right?
* note the comments in exercise/cache/points.py about a potential bug

skulonen commented on exercise/static/exercise/assessment.js:
```diff
@@ -0,0 +1,98 @@
+  feedback2.addClass('feedback-toggle hide');
+  var label1 = feedback1.find('label');
+  var label2 = feedback2.find('label');
+  var feedbackToggleLink = $('<a href="#"></a>')
```
> For actions that modify the current page content like switching tabs or hiding a section, use <button>

Bootstrap uses `<a>` elements for tabs (and "pills"). If we change them into buttons because we would then lose Bootstrap's default styles for tabs/pills.

The feedback toggle link that you pointed out here _can_ be changed to a button since it's not a tab element, but there's a pill element in _assessment_panel.html that is used for switching between grader feedback and grader errors. What should I do about that? I guess I could turn it into buttons and use the new visibility switching functionality instead of Bootstrap's pills.

skulonen replied:
The "close sidebar" link is also a pill element. I could turn it into a button, but then it would stick out like sore thumb:
![image](https://user-images.githubusercontent.com/28843787/135863350-da798ca5-c48f-4063-b686-9259798b2ce2.png)
Or maybe I could make the button smaller and put it next to the language switch button?

skulonen replied:
> Or maybe I could make the button smaller and put it next to the language switch button?

![image](https://user-images.githubusercontent.com/28843787/135990875-c76e3b2f-be44-44ca-8bdd-7780af9b3dfe.png)
Looks pretty good in my opinion. I also managed to make the sidebar expander element a button without altering its appearance.

Mankro replied:
Sorry, I didn't realize that Bootstrap 3 has this limitation. We don't need to make this too complicated. If it is not trivial (and quick to finish) to change `<a>` to `<button>`, then don't spend time on it. If you already did something, that's fine too. If you need to use `<a>` for an action (that is not a link to another page), then it should have `role="button"`.

The screenshot of the new "close sidebar" button looks good to me.

skulonen replied:
Fixed, the feedback toggles are now buttons.

skulonen commented on course/templates/course/course_base.html:
```diff
@@ -32,9 +32,20 @@
+				<li>
+					<a href="#" id="course-sidebar-collapser" role="button">
+						<span class="glyphicon glyphicon-chevron-left" aria-hidden="true"></span>
+						<span>{% trans "CLOSE_SIDEBAR" %}</span>
```
Fixed

skulonen commented on assets/js/aplus.js:
```diff
@@ -118,6 +119,31 @@ $(function() {
     $(window).on('scroll', modifyMenu);
     $(window).on('resize', updateMenu);
 
+    var setSidebarState = function(collapsed) {
```
> I don't know any reason why var x = function(){} has been used instead of function x() {}

Me neither. I changed all of my function variables into standard functions.

skulonen commented on assets/js/resize.js:
```diff
@@ -0,0 +1,113 @@
+
+/**
+* This plugin makes the children of an element resizable horizontally. It
+* it creates "handle" elements between the children that the user can click and
```
Fixed

skulonen commented on assets/sass/legacy/_main.scss:
```diff
@@ -198,6 +203,34 @@ table.src tr td.src {
 #course-content a{
 	@include focus-with-background-and-color;
 }
+#course-sidebar-expander {
```
I extracted the sidebar styles and the styles related to code highlighting into their own components.

skulonen commented on exercise/staff_views.py:
```diff
@@ -146,25 +175,100 @@ def form_valid(self, form):
-        #    message=note,
-        #)
+        # Set other submissions as not final if this one is final.
+        if self.submission.force_exercise_points:
```
Fixed

skulonen commented on exercise/staff_views.py:
```diff
@@ -146,25 +175,100 @@ def form_valid(self, form):
+    def get(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
+        # Query submitters who have not been assessed yet.
+        submitter = None
+        submitters = UserProfile.objects\
```
Fixed

skulonen commented on exercise/staff_views.py:
```diff
@@ -146,25 +175,100 @@ def form_valid(self, form):
+        # Find the submitter's best submission using the cache.
+        cache = CachedPoints(self.instance, submitter.user, self.content)
+        ids = cache.submission_ids(exercise_id=self.exercise.id, best=True)
+        if len(ids) == 0:
```
Fixed

skulonen commented on exercise/static/exercise/assessment.js:
```diff
@@ -0,0 +1,98 @@
+  var feedback1 = $('#id_assistant_feedback').closest('.form-group');
+  var feedback2 = $('#id_feedback').closest('.form-group');
+  feedback1.addClass('feedback-toggle');
+  feedback2.addClass('feedback-toggle hide');
```
Fixed

skulonen commented on locale/en/LC_MESSAGES/django.po:
```diff
@@ -1019,23 +1017,15 @@ msgstr "You can change the language from your account page."
 msgstr "Breadcrumb"
 
-#: course/templates/course/course_base.html
-msgid "UNENROLL_WARNING_DESCRIPTION"
```
Yes, looks like I accidentally committed them into this branch even though they're not related to this PR. I reverted the change.

skulonen commented on exercise/templates/exercise/staff/_submissions_table.html:
```diff
@@ -6,11 +6,15 @@
 
 <div class="clearfix">
 	<div class="pull-right">
+		<a class="aplus-button--secondary aplus-button--xs" role="button" href="{{ exercise|url:'submission-next-unassessed' }}">
```
Fixed

skulonen commented on exercise/templates/exercise/staff/_submission_data_modal.html:
```diff
@@ -0,0 +1,38 @@
+	<div class="modal-dialog" role="document">
+		<div class="modal-content">
+			<div class="modal-header">
+				<button type="button" class="close" data-dismiss="modal" aria-label="Close">
```
Fixed

skulonen commented on exercise/templates/exercise/staff/_assessment_panel.html:
```diff
@@ -0,0 +1,145 @@
+<div class="panel panel-default assessment-panel">
+	<div class="panel-heading">
+		{# Heading: default mode #}
+		<div class="assessment-bar assessment-toggle{% if form.errors %} hide{% endif %}">
```
Fixed

skulonen commented on exercise/templates/exercise/staff/_assessment_panel.html:
```diff
@@ -0,0 +1,145 @@
+		</div>
+
+		{# Heading: assessment mode #}
+		<form class="assessment-bar assessment-toggle{% if not form.errors %} hide{% endif %}" method="post" role="form">
```
Fixed

skulonen commented on exercise/templates/exercise/staff/inspect_submission.html:
```diff
@@ -14,68 +15,93 @@
+			<h4>
+				{% translate "SUBMITTERS" %}
+				{% if is_teacher %}
+				<a href="{{ submission|url:'submission-edit-submitters' }}" class="aplus-button--secondary aplus-button--xs" role="button">
```
Fixed

skulonen commented on exercise/templates/exercise/staff/inspect_submission.html:
```diff
@@ -14,68 +15,93 @@
+								title="{% translate 'ADD_ONE_EXTRA_SUBMISSION_FOR_STUDENT_IN_EXERCISE' %}"
+							>
+								{% csrf_token %}
+								<input class="aplus-button--secondary aplus-button--xs" type="submit" value="+1" />
```
Fixed. I moved the badges and buttons out of the h4 element for both of the heading elements on this page.

skulonen commented on exercise/templates/exercise/staff/_submissions_table_compact.html:
```diff
@@ -0,0 +1,77 @@
+		{% for s in submissions reversed %}
+		<tr
+			{% if s != submission %}
+			class="submissions-toggle hide"
```
Fixed

skulonen commented on exercise/templates/exercise/staff/_submitters_table.html:
```diff
@@ -0,0 +1,96 @@
+	</div>
+	<p>
+		{% exercise_text_stats exercise %}
+		<a class="aplus-button--secondary aplus-button--xs" role="button" href="{{ exercise|url:'submission-list' }}">
```
Fixed

skulonen commented on exercise/cache/points.py:
```diff
@@ -113,23 +123,36 @@ def r_augment(children):
+                        'unofficial': False,
                     })
+                    forced_points = True
+                if not forced_points:
```
Fixed. The `foced_points` flag is now stored in the `entry` instead of a global flag.

skulonen commented:
@Mankro I went through your comments and made changes where necessary. I pushed the changes in 28 small commits so that I can easily squash them with the appropriate main commits when cleaning the commit history.

I also implemented the ability to filter submitters by grade on the submitter list page.

>perhaps later in the autumn, we could refactor the points code so that similar logic is not repeated in many places (DRY principle), but not now due to time constraints.

Yeah, especially now that I added the previously mentioned grade filtering, which is now yet another place where points are calculated. And when the upcoming delayed feedback PR is merged, the points calculation becomes even more complicated...

We should figure out a way to unify these calculations. The points cache is almost ideal because it is cached, but it has one problem: each CachedPoints instance is related to a single student, and as such it can't be used nicely on pages where the points of multiple users are listed (submitter list, API).

>you didn't use const in JavaScript code, but it would be preferred whenever the variable is not supposed to be reassigned. Perhaps it is not worth the time to go over all your JS code now, though?

I fixed this too. Const is now used where appropriate in the code that I edited in this PR.

Mankro commented:
>>    submitted file line numbers are sometimes messed up like in #794. In my test, the line numbers worked temporarily when the submission was resubmitted to auto-grading and the feedback was replaced with the "waiting for results" feedback.

>Once again, I can't reproduce this... Could you tell me what browser and which exercise you used? And also what file did you submit?

Exercise: simple Python exercise in the test-course
https://github.com/apluslms/test-course/blob/multilang/links1/chapter1_en.rst

Submitted file: functions.py
[functions.py.txt](https://github.com/apluslms/a-plus/files/7300191/functions.py.txt)

I was using Firefox on Ubuntu 20.04.

---------------------------
As for the other questions, OK!

skulonen commented:
I have now pushed fixes to these new issues (still couldn't reproduce the line number issue, but that one has its own Github issue anyway). I'll start working on cleaning the commit history, but feel free to comment if there's anything else you think should be fixed.
